fcrepo-search-impl/src/main/java/org/fcrepo/search/impl/DbSearchIndexImpl.java

```diff
@@ -346,33 +363,69 @@ private void deleteRdfTypeAssociations(final Long resourceId) {
                 deleteParams);
     }
 
-    private ArrayList<Long> findOrCreateRdfTypesInDb(final List<URI> rdfTypes) {
-        final var jdbcInsertRdfTypes = new SimpleJdbcInsert(this.jdbcTemplate.getJdbcTemplate());
-        jdbcInsertRdfTypes.withTableName(SEARCH_RDF_TYPE_TABLE).usingGeneratedKeyColumns(
-                ID_COLUMN);
-        final var rdfTypeIds = new ArrayList<Long>();
-        for (var rdfTypeUri : rdfTypes) {
-            final var typeParams = new MapSqlParameterSource();
-            typeParams.addValue(RDF_TYPE_URI_PARAM, rdfTypeUri.toString());
-            final var results = jdbcTemplate.queryForList(SELECT_RDF_TYPE_ID,
-                    typeParams);
-            if (CollectionUtils.isEmpty(results)) {
-                final Number key = jdbcInsertRdfTypes.executeAndReturnKey(typeParams);
-                rdfTypeIds.add(key.longValue());
-            } else {
-                rdfTypeIds.add((long) results.get(0).get(ID_COLUMN));
+    private List<Long> findOrCreateRdfTypesInDb(final List<URI> rdfTypes) {
+        final List<String> rdfTypes_str = rdfTypes.stream().map(URI::toString).collect(Collectors.toList());
+
+        final List<RdfType> results = jdbcTemplate.query(SELECT_RDF_TYPE_ID,
+                Map.of(RDF_TYPE_URI_PARAM, rdfTypes_str), RDF_TYPE_ROW_MAPPER);
+        // List of existing type ids.
+        final var rdfTypeIds = results.stream().map(RdfType::getTypeId).collect(Collectors.toList());
+        // List of existing type uris.
+        final var rdfTypeUris = results.stream().map(RdfType::getTypeUri).collect(Collectors.toList());
+        // Type uris that don't already have a record. Needs to be a set to avoid inserting the same URI and
+        final var missingUris = rdfTypes_str.stream().filter(t -> !rdfTypeUris.contains(t))
+                .collect(Collectors.toSet());
+
+        if (!missingUris.isEmpty()) {
+            final List<MapSqlParameterSource> parameterSourcesList = new ArrayList<>();
+            missingUris.forEach(u -> {
+                final var assocParams = new MapSqlParameterSource();
+                assocParams.addValue(RDF_TYPE_URI_PARAM, u);
+                LOGGER.debug("Adding rdf type uri: " + u);
+                parameterSourcesList.add(assocParams);
+            });
+            final MapSqlParameterSource[] psArray = parameterSourcesList.toArray(new MapSqlParameterSource[0]);
+            // Batch insert all the records.
+            jdbcTemplate.batchUpdate(INSERT_RDF_TYPE, psArray);
```

Comment:
As you noted, this can fail if multiple threads attempt to concurrently add the same types. Does it not throw an exception in that case? If so, I think you could safely swallow it and proceed because you'll get a failure on line 394 if the types weren't actually created.

Reply:
What happens if there are two concurrent requests, one is inserting types A, B, and C and the other is inserting B, C, and D? Since it's a batch update, will the end result be that the database either contains A, B, and C or B, C, and D? Or will it contain all of the types?

Reply:
It looks like we need to go back and use the standard jdbc batch which should return a BatchUpdateException and return success and failure counts. I'll give this a try next